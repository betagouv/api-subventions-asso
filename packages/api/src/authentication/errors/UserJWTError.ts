import HttpError from "../../shared/errors/httpErrors/HttpError";

export default class UserJWTError extends HttpError {
    status = 401
}