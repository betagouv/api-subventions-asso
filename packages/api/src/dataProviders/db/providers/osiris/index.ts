import osirisRequestPort from "./osiris.request.port";
import osirisActionPort from "./osiris.action.port";

export { osirisRequestPort, osirisActionPort };
