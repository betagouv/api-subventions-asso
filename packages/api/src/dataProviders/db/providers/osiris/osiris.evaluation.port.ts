import { FindOneAndUpdateOptions } from "mongodb";
import OsirisEvaluationEntity from "../../../../modules/providers/osiris/entities/OsirisEvaluationEntity";
import MongoRepository from "../../../../shared/MongoRepository";
import Siret from "../../../../valueObjects/Siret";
import Siren from "../../../../valueObjects/Siren";

export class OsirisEvaluationPort extends MongoRepository<OsirisEvaluationEntity> {
    readonly collectionName = "osiris-evaluation";

    public async findByActionId(actionId: string) {
        return this.collection.findOne({
            "indexedInformations.osirisActionId": actionId,
        });
    }

    public async add(request: OsirisEvaluationEntity) {
        await this.collection.insertOne(request);
        return request;
    }

    public async update(request: OsirisEvaluationEntity) {
        const options = { returnNewDocument: true, includeResultMetadata: true } as FindOneAndUpdateOptions;
        const { _id, ...requestWithoutId } = request;
        return (await this.collection.findOneAndUpdate(
            { "indexedInformations.osirisActionId": request.indexedInformations.osirisActionId },
            { $set: requestWithoutId },
            options,
            //@ts-expect-error -- mongo typing expects no metadata
        ))!.value as OsirisEvaluationEntity;
    }

    public findsBySiret(siret: Siret) {
        return this.collection
            .find({
                "legalInformations.siret": siret.value,
            })
            .toArray();
    }

    public findBySiren(siren: Siren) {
        return this.collection
            .find({
                "legalInformations.siret": new RegExp(`^${siren.value}\\d{5}`),
            })
            .toArray();
    }

    public cursorFind(query = {}) {
        return this.collection.find(query);
    }

    async createIndexes() {
        await this.collection.createIndex({ "indexedInformations.osirisActionId": 1 }, { unique: true });
        await this.collection.createIndex({ "legalInformations.siret": 1 });
    }
}

const osirisEvaluationPort = new OsirisEvaluationPort();

export default osirisEvaluationPort;
