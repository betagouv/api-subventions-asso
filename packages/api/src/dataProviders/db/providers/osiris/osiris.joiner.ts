import { AggregationCursor } from "mongodb";
import db from "../../../../shared/MongoConnection";
import osirisActionPort from "./osiris.action.port";
import osirisRequestPort from "./osiris.request.port";
import OsirisRequestEntity from "../../../../modules/providers/osiris/entities/OsirisRequestEntity";
import OsirisActionEntity from "../../../../modules/providers/osiris/entities/OsirisActionEntity";

export type OsirisRequestWithActions = OsirisRequestEntity & { actions: OsirisActionEntity[] };

export class OsirisJoiner {
    applicationCollection = db.collection(osirisRequestPort.collectionName);

    private get joinPipeline() {
        return [
            {
                $lookup: {
                    from: osirisActionPort.collectionName,
                    localField: osirisRequestPort.joinIndexes.osirisActionPort,
                    foreignField: osirisActionPort.joinIndexes.osirisRequestPort,
                    as: "actions",
                },
            },
        ];
    }

    findAllCursor(): AggregationCursor<OsirisRequestWithActions> {
        return this.applicationCollection.aggregate(this.joinPipeline);
    }

    findByExerciseCursor(exercise: number): AggregationCursor<OsirisRequestWithActions> {
        return this.applicationCollection.aggregate([
            { $match: { "providerInformations.exercise": exercise } },
            ...this.joinPipeline,
        ]);
    }
}

const osirisJoiner = new OsirisJoiner();
export default osirisJoiner;
