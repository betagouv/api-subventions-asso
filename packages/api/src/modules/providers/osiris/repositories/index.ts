import osirisRequestRepository from "./osiris.request.repository";
import osirisActionRepository from "./osiris.action.repository";
import osirisEvaluationRepository from "./osiris.evaluation.repository";

export { osirisRequestRepository, osirisActionRepository, osirisEvaluationRepository };
