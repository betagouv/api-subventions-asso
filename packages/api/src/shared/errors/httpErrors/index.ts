export * from "./BadRequestError";
export * from "./InternalServerError";
export * from "./NotFoundErrors";
export * from "./ConflictError";
export * from "./ForbiddenError";
export * from "./UnauthorizedError";
