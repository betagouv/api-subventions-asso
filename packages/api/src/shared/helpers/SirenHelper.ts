import { Siren, Siret } from "dto";

export function siretToSiren(siret: Siret): Siren {
    return siret.slice(0, 9);
}

export function siretToNIC(siret: Siret) {
    return siret.slice(9, 14);
}
