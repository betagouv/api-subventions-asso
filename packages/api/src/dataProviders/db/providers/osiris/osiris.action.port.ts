import { FindOneAndUpdateOptions } from "mongodb";
import { MongoCnxError } from "../../../../shared/errors/MongoCnxError";
import OsirisActionEntity from "../../../../modules/providers/osiris/entities/OsirisActionEntity";
import OsirisActionEntityDbo from "../../../../modules/providers/osiris/entities/OsirisActionEntityDbo";
import MongoPort from "../../../../shared/MongoPort";
import Siren from "../../../../valueObjects/Siren";
import OsirisActionAdapter from "./osirisAction.adapter";

export class OsirisActionPort extends MongoPort<OsirisActionEntityDbo> {
    collectionName = "osiris-actions";

    async createIndexes() {
        await this.collection.createIndex({ "indexedInformations.uniqueId": 1 }, { unique: true });
        await this.collection.createIndex({ "indexedInformations.osirisActionId": 1 });
        await this.collection.createIndex({ "indexedInformations.requestUniqueId": 1 });
        await this.collection.createIndex({ "indexedInformations.compteAssoId": 1 });
        await this.collection.createIndex({ "indexedInformations.siret": 1 });
    }

    // Action Part
    public async add(osirisAction: OsirisActionEntity) {
        await this.collection.insertOne(OsirisActionAdapter.toDbo(osirisAction));
        return osirisAction;
    }

    /*
     * @deprecated
     * */
    public async update(osirisAction: OsirisActionEntity) {
        const options: FindOneAndUpdateOptions = { returnDocument: "after", includeResultMetadata: true };
        const { _id, ...actionWithoutId } = OsirisActionAdapter.toDbo(osirisAction);
        const updateRes = await this.collection.findOneAndUpdate(
            { "indexedInformations.uniqueId": osirisAction.indexedInformations.uniqueId },
            { $set: actionWithoutId },
            options,
        );

        //@ts-expect-error -- mongo typing expects no metadata
        const dbo = updateRes?.value;
        if (!dbo) throw new MongoCnxError();
        return OsirisActionAdapter.toEntity(dbo);
    }

    public upsertOne(osirisAction: OsirisActionEntity) {
        const options = { upsert: true } as FindOneAndUpdateOptions;
        const { _id, ...actionWithoutId } = osirisAction;
        return this.collection.updateOne(
            { "indexedInformations.uniqueId": osirisAction.indexedInformations.uniqueId },
            { $set: actionWithoutId },
            options,
        );
    }

    public async bulkUpsert(osirisActions: OsirisActionEntity[]) {
        const bulk = osirisActions.map(a => {
            const { _id, ...actionWithoutId } = a;
            return {
                updateOne: {
                    filter: { "indexedInformations.uniqueId": a.indexedInformations.uniqueId },
                    update: { $set: actionWithoutId },
                    upsert: true,
                },
            };
        });
        return bulk.length ? this.collection.bulkWrite(bulk, { ordered: false }) : Promise.resolve();
    }

    public async findByUniqueId(uniqueId: string) {
        const dbo = await this.collection.findOne({
            "indexedInformations.uniqueId": uniqueId,
        });
        if (!dbo) return null;
        return OsirisActionAdapter.toEntity(dbo);
    }

    public cursorFind(query = {}) {
        return this.collection.find(query);
    }

    public async findByRequestUniqueId(requestUniqueId: string) {
        const dbos = await this.collection.find({ "indexedInformations.requestUniqueId": requestUniqueId }).toArray();
        return dbos.map(dbo => OsirisActionAdapter.toEntity(dbo));
    }

    public async findBySiren(siren: Siren) {
        const dbos = await this.collection
            .find({ "indexedInformations.siret": new RegExp(`^${siren.value}\\d{5}`) })
            .toArray();
        return dbos.map(dbo => OsirisActionAdapter.toEntity(dbo));
    }
}

const osirisActionPort: OsirisActionPort = new OsirisActionPort();

export default osirisActionPort;
