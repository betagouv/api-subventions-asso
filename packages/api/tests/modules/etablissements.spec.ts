import request from "supertest";
import { createAndGetAdminToken, createAndGetUserToken } from "../__helpers__/tokenHelper";
import etablissementService from "../../src/modules/etablissements/etablissements.service";
import statsService from "../../src/modules/stats/stats.service";
import { siretToSiren } from "../../src/shared/helpers/SirenHelper";
import associationsService from "../../src/modules/associations/associations.service";
import { BadRequestError } from "../../src/shared/errors/httpErrors";
import OsirisRequestEntityFixture from "./providers/osiris/__fixtures__/entity";
import { osirisRequestRepository } from "../../src/modules/providers/osiris/repositories";
import { versionnedUrl } from "../__helpers__/routeHelper";

const g = global as unknown as { app: unknown };

const ETABLISSEMENT_SIRET = "12345678901234";

describe("/etablissement", () => {
    const getSubventionsMock: jest.SpyInstance = jest.spyOn(etablissementService, "getSubventions");
    afterAll(() => {
        getSubventionsMock.mockRestore();
    });

    describe("/{structure_identifier}/subventions", () => {
        describe("on success", () => {
            const SUBVENTIONS = ["subventions"];
            const SUBVENTION_FLUX = [{ subventions: SUBVENTIONS }];
            beforeEach(() => {
                getSubventionsMock.mockImplementationOnce(() => ({
                    toPromise: async () => SUBVENTION_FLUX,
                }));
            });
            it("should return 200", async () => {
                const actual = (
                    await request(g.app)
                        .get(versionnedUrl(`/etablissement/${ETABLISSEMENT_SIRET}/subventions`))
                        .set("x-access-token", await createAndGetUserToken())
                        .set("Accept", "application/json")
                ).statusCode;

                expect(actual).toBe(200);
            });

            it("should return an object with subventions", async () => {
                const expected = { subventions: SUBVENTIONS };
                const actual = (
                    await request(g.app)
                        .get(versionnedUrl(`/etablissement/${ETABLISSEMENT_SIRET}/subventions`))
                        .set("x-access-token", await createAndGetUserToken())
                        .set("Accept", "application/json")
                ).body;

                expect(actual).toEqual(expected);
            });
        });
    });

    describe("/{structure_identifier}", () => {
        beforeEach(async () => {
            await osirisRequestRepository.add(OsirisRequestEntityFixture);
        });
        it("should add one visits on stats AssociationsVisit", async () => {
            const beforeRequestTime = new Date();
            const a = await request(g.app)
                .get(versionnedUrl(`/etablissement/${OsirisRequestEntityFixture.legalInformations.siret}`))
                .set("x-access-token", await createAndGetUserToken())
                .set("Accept", "application/json");

            const actual = await statsService.getTopAssociationsByPeriod(1, beforeRequestTime, new Date());
            const expected = [
                {
                    name: siretToSiren(OsirisRequestEntityFixture.legalInformations.siret),
                    visits: 1,
                },
            ];

            expect(actual).toEqual(expected);
        });

        it("should not add one visits on stats AssociationsVisit beacause user is admin", async () => {
            const beforeRequestTime = new Date();
            await request(g.app)
                .get(versionnedUrl(`/etablissement/${OsirisRequestEntityFixture.legalInformations.siret}`))
                .set("x-access-token", await createAndGetAdminToken())
                .set("Accept", "application/json");
            const actual = await statsService.getTopAssociationsByPeriod(1, beforeRequestTime, new Date());

            expect(actual).toHaveLength(0);
        });

        it("should not add one visits on stats AssociationsVisit beacause user is not authentified", async () => {
            const beforeRequestTime = new Date();
            await request(g.app)
                .get(versionnedUrl(`/etablissement/${OsirisRequestEntityFixture.legalInformations.siret}`))
                .set("Accept", "application/json");
            const actual = await statsService.getTopAssociationsByPeriod(1, beforeRequestTime, new Date());

            expect(actual).toHaveLength(0);
        });

        it("should not add one visits on stats AssociationsVisit beacause status is not 200", async () => {
            const beforeRequestTime = new Date();
            jest.spyOn(associationsService, "getAssociation").mockImplementationOnce(() => {
                throw new BadRequestError();
            });
            await request(g.app)
                .get(versionnedUrl(`/etablissement/${OsirisRequestEntityFixture.legalInformations.siret}`))
                .set("Accept", "application/json");
            const actual = await statsService.getTopAssociationsByPeriod(1, beforeRequestTime, new Date());

            expect(actual).toHaveLength(0);
        });
    });
});
