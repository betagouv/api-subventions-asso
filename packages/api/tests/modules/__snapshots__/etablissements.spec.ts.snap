// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/etablissement /siret/grants should return grants 1`] = `
Object {
  "subventions": Array [],
}
`;
